ansible/inventory/multi_inventory.py

```diff
@@ -5,29 +5,137 @@
+        if self.config.has_key('cluster_vars') and \
+            self.config['cluster_vars'].has_key('clusters'):
+            self.clusters = self.config['cluster_vars']['clusters'].keys()
+
+
+    @property
+    def account_data(self):
+        """property for account_data"""
+        if self._account_data == None:
+            self._account_data = MultiInventoryUtils.get_inventory_from_cache(os.path.join(self.cache_path, self.name))
+
+        return self._account_data
+
+
+    @account_data.setter
+    def account_data(self, data):
+        """setter for account_data"""
+        self._account_data = data
+
+    def run_provider(self, force=False, valid_cache=False):
+        """Setup the provider call with proper variables and call self.get_provider_tags
+
+           force:  Whether to force an update on the account
+           valid_cache:  Whether or not to validate if is_cache_valid.
+                         This happens in case of an account update
+
+        """
+        # force an update??
+        if not force:
+            # check if we want a valid cache or if the cache is valid
+            if not valid_cache or \
+               MultiInventoryUtils.is_cache_valid(os.path.join(self.cache_path, self.name), self.cache_age):
+                return {'cached': True, 'data': self.account_data}
+
+        if self.config.has_key('provider_files'):
+            tmp_dir = self.generate_config()
+
+	# Update env vars after creating provider_config_files
+	# so that we can grab the tmp_dir if it exists
+        env = self.config.get('env_vars', {})
+        if env and tmp_dir:
+            for key, value in env.items():
+                env[key] = Template(value).substitute(tmpdir=tmp_dir)
+
+        if not env:
+            env = os.environ
+
+
+        provider = self.config['provider']
+        # Allow relatively path'd providers in config file
+        if os.path.isfile(os.path.join(FILE_PATH, self.config['provider'])):
+            provider = os.path.join(FILE_PATH, self.config['provider'])
+
+        # check to see if provider exists
+        if not os.path.isfile(provider) or not os.access(provider, os.X_OK):
+            raise RuntimeError("Problem with the provider.  Please check path " \
+                        "and that it is executable. (%s)" % provider)
+
+        cmds = [provider, '--list']
+
+        if 'aws' in provider.lower():
+            cmds.append('--refresh-cache')
+
+        return {'cached': False,
+                'proc': subprocess.Popen(cmds, stderr=subprocess.PIPE, stdout=subprocess.PIPE, env=env)}
+
+    def generate_config(self):
+        """Generate the provider_files in a temporary directory"""
+        prefix = 'multi_inventory.'
+        tmp_dir_path = tempfile.mkdtemp(prefix=prefix)
+        atexit.register(MultiInventoryUtils.cleanup, [tmp_dir_path])
+        for provider_file in self.config['provider_files']:
+            filedes = open(os.path.join(tmp_dir_path, provider_file['name']), 'w+')
```

Comment:
I think this will fail to close the file descriptor if there's an exception in the call to Template / substitute or in writing to the file.

I understand that these are temp files, so probably not the end of the world, but still good practice.

Consider using `with` like this:
```python
with open(os.path.join(tmp_dir_path, provider_file['name']), 'w+') as filedes:
    content = Template(provider_file['contents']).substitute(tmpdir=tmp_dir_path)
    filedes.write(content)
```